File: packages/models/src/committee/CommitteePosition.ts
import type { LanguageCode } from '../util/Language'
import type { Committee } from './Committee'
import type { CommitteePositionCategory } from './CommitteeCategory'

/**
 * @interface CommitteePosition
 * @description Committee position model
 *
 * @param {string} email - Committee position email
 * @param {string} title - Committee position title
 * @param {string} description - Committee position description
 * @param {'ADMIN' | 'BOARD' | 'COMMITTEE'} role - Committee position role
 * @param {boolean} active - Whether the position is active or not
 * @param {number} weight - Committee position weight
 */
export interface CommitteePosition {
  committee_position_id: string
  author_type: 'COMMITTEE_POSITION'
  email?: string
  role: 'ADMIN' | 'BOARD' | 'COMMITTEE'
  active: boolean
  weight: number
  base: boolean
  category: CommitteePositionCategory
  committee?: Committee
  translations: CommitteePositionTranslation[]
}

export interface CommitteePositionTranslation {
  title: string
  description: string
  language_code: LanguageCode
}

export interface CommitteePositionRecruitment {
  committee_position: CommitteePosition
  start_date: string
  end_date: string
  translations: [
    {
      description: string
      link_url: string
    },
  ]
}

File: apps/backend/core-api/dto/committees/committee_position.py
import uuid

import msgspec

from dto.committees.committee import CommitteeDTO
from models.committees.committee_position import (
    CommitteePositionCategory,
    CommitteePositionsRole,
)


class CommitteePositionTranslationDTO(msgspec.Struct):
    title: str
    description: str
    language_code: str


class PublicCommitteePositionDTO(msgspec.Struct):
    email: str | None
    translations: list[CommitteePositionTranslationDTO]
    committee_id: uuid.UUID

    @classmethod
    def from_orm_with_language(
        cls, obj, language_code: str
    ) -> "PublicCommitteePositionDTO":
        """
        Create a PublicCommitteePositionDTO from an ORM object and a language code.
        """
        # Filter translations by the specified language code
        translations = [
            translation
            for translation in obj.translations
            if translation.language_code == language_code
        ]

        # Create a new PublicCommitteePositionDTO instance with the filtered translations
        return cls(
            email=obj.email,
            translations=translations,
            committee_id=obj.committee_id,
        )


class PublicCommitteePositionDTOWithCommitteeDTO(
    PublicCommitteePositionDTO,
):
    committee: CommitteeDTO


class ProtectedCommitteePositionDTO(PublicCommitteePositionDTO):
    weight: int
    role: CommitteePositionsRole
    active: bool
    category: CommitteePositionCategory | None
    base: bool


class ProtectedCommitteePositionDTOWithCommitteeDTO(
    ProtectedCommitteePositionDTO,
):
    committee: CommitteeDTO


class CommitteePositionRecruitmentTranslationDTO(msgspec.Struct):
    description: str
    link_url: str
    language_code: str


class CommitteePositionRecruitmentDTO(msgspec.Struct):
    start_date: str
    end_date: str
    committee_position: PublicCommitteePositionDTO
    translations: list[CommitteePositionRecruitmentTranslationDTO]

    @classmethod
    def from_orm_with_language(
        cls, obj, language_code: str
    ) -> "CommitteePositionRecruitmentDTO":
        """
        Create a CommitteePositionRecruitmentDTO from an ORM object and a language code.
        """
        # Filter translations by the specified language code
        translations = [
            translation
            for translation in obj.translations
            if translation.language_code == language_code
        ]

        # Create a new CommitteePositionRecruitmentDTO instance with the filtered translations
        return cls(
            start_date=obj.start_date,
            end_date=obj.end_date,
            committee_position=PublicCommitteePositionDTO.from_orm_with_language(
                obj.committee_position, language_code
            ),
            translations=translations,
        )
